components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -6,82 +6,406 @@
+  pref_change_registrar_.Init(local_pref_);
+  pref_change_registrar_.Add(kReferralPromoCode,
+      base::BindRepeating(&NTPBackgroundImagesService::OnPreferenceChanged,
+      base::Unretained(this)));
+}
+
+bool NTPBackgroundImagesService::IsAlreadyRegistered(
+    const std::string& component_id) {
+  std::vector<std::string> component_ids;
+  if (!component_update_service_)
+    return false;
+
+  component_ids = component_update_service_->GetComponentIDs();
+  return std::find(component_ids.begin(),
+                   component_ids.end(),
+                   component_id) != component_ids.end();
+}
+
+void NTPBackgroundImagesService::UnRegisterSponsoredImagesComponentIfRunning() {
+  const std::string locale =
+      brave_ads::LocaleHelper::GetInstance()->GetLocale();
+  const auto& data = GetSponsoredImagesComponentData(
+      brave_ads::LocaleHelper::GetCountryCode(locale));
+  if (!data)
+    return;
+
+  if (!component_update_service_)
+    return;
+
+  if (IsAlreadyRegistered(data->component_id)) {
+    component_update_service_->UnregisterComponent(data->component_id);
+    DVLOG(2) << __func__ << ": Unregister SI Component.";
+  }
+}
+
+void NTPBackgroundImagesService::StartSuperReferrerComponent(
+    const std::string& super_referral_code) {
+  DCHECK(mapping_table_value_.is_dict());
+
+  auto* value = mapping_table_value_.FindDictKey(super_referral_code);
+  if (!value) {
+    NOTREACHED();
     return;
   }
 
-  // Early return for test.
-  if (!cus)
+  is_super_referrer_active_ = true;
+  if (IsAlreadyRegistered(*value->FindStringKey(kComponentID)))
     return;
 
+  local_pref_->Set(prefs::kNewTabPageCachedSuperReferrerComponentInfo, *value);
+
+  UnRegisterSponsoredImagesComponentIfRunning();
+
+  DVLOG(2) << __func__ << ": Start NTP SR component";
+  DCHECK(IsValidSuperReferrerComponentInfo(*value));
+  RegisterNTPBackgroundImagesComponent(
+      component_update_service_,
+      *value->FindStringKey(kPublicKey),
+      *value->FindStringKey(kComponentID),
+      base::StringPrintf("NTP Super Referrer (%s)",
+                         value->FindStringKey(kCompanyName)->c_str()),
+      base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
+                          weak_factory_.GetWeakPtr(),
+                          true));
+}
+
+void NTPBackgroundImagesService::StartSponsoredImagesComponent() {
   const std::string locale =
       brave_ads::LocaleHelper::GetInstance()->GetLocale();
-  if (const auto& data = GetRegionalComponentData(
-          brave_ads::LocaleHelper::GetCountryCode(locale))) {
-    RegisterNTPBackgroundImagesComponent(cus, data.value(),
-        base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
-                            weak_factory_.GetWeakPtr()));
+  const auto& data = GetSponsoredImagesComponentData(
+      brave_ads::LocaleHelper::GetCountryCode(locale));
+  if (!data)
+    return;
+
+  is_super_referrer_active_ = false;
```

Comment:
why do we have a flag only for super_referrer? That logic is not going to work when regular background images are also integrated into this. If you want to know which component type is currently active then I think it makes more sense to check the loaded component_id

Reply:
Using |is_super_referrer_active_| was inevitable.
I wrote about why this is needed at https://github.com/brave/brave-core/pull/4866/files#diff-992d3e514ddb745d42ac450231255edfR440.

Regarding to regular background images component, I think that component should be registered independently with other SI or SR because it should be available always.